README.md
# apostrophe-site-map

This module generates XML and plaintext sitemaps for sites powered by the [Apostrophe](https://apostrophenow.org) CMS.

It serves two purposes: [white-hat SEO](https://support.google.com/webmasters/answer/183668?hl=en&ref_topic=6080646&rd=1) and content strategy.

## SEO with sitemaps

A frequently updated and accurate XML sitemap allows search engines to index your content more quickly and spot new pages immediately. But *an out-of-date sitemap is worse than nothing and will damage your site's SEO.*

This module generates a sitemap that includes all of the pages on your site that are visible to the public, including "pieces" such as events, and blog posts. And it does so dynamically, with a short cache lifetime, so your sitemap is not out of date.

## How to use it

* Install the module.

`npm install --save apostrophe-site-map`

* Configure it in `app.js`, as one of your modules.

```javascript
{
  // You should configure `baseUrl` to ensure full URLs in your sitemap
  baseUrl: 'http://example.com',
  modules: {
    'apostrophe-site-map': {
      // array of doc types you do NOT want
      // to include, even though they are
      // accessible on the site. You can also
      // do this at the command line.
      excludeTypes: []
    }
  }
}
```

#### Alternative configuration

If you don't like to modify/overwrite the baseUrl for the site or keep the site without a baseUrl, you can add baseUrl in the configuration of the module:

```javascript
{
  // No baseUrl here
  modules: {
    'apostrophe-site-map': {
      baseUrl: 'http://example.com',
      excludeTypes: []
    }
  }
}
```

* Just launch your site as you normally would. In development that might just be:

```
node app
```

* Access `http://localhost:3000/sitemap.xml` (in production, of course, the hostname is different).

> **AN IMPORTANT WARNING: if you ALREADY have a STATIC public/sitemap.xml file, THAT FILE WILL BE SENT INSTEAD.** Remove it. Also, SITEMAPS ARE CACHED for one hour by default, so you won't see changes instantly. Read on for how to change the cache lifetime, and what you can realistically expect from Google.

### Clearing the cache, and changing the cache lifetime

To better support multiple-server environments, this module now serves sitemaps directly and caches them in your database. That way we don't have to worry about whether a static file exists in a given environment, running the same task on multiple servers, etc.

By default sitemaps are cached for 1 hour. You can change this by specifying the `cacheLifetime` option to this module, in seconds. However, don't get too excited: Google [usually does not check a sitemap more often than a few times a month](https://webmasters.stackexchange.com/questions/43874/how-often-does-gwt-check-dynamic-sitemaps).

You can clear the cache at any time with this command line task:

```
node app apostrophe-site-map:clear
```

This will force a new sitemap to be generated on the next request.

### Generating the sitemap ahead of time

You can use this command line task to update the sitemap in Apostrophe's cache at any time, rather than waiting for it to expire after an hour and generate again on the next request:

```
node app apostrophe-site-map:map --update-cache
```

If your site has many pages and pieces, generating the sitemap dynamically may take a long time. Scheduling the above task to run at least twice an hour via a [cron job](https://www.howtogeek.com/101288/how-to-schedule-tasks-on-linux-an-introduction-to-crontab-files/) guarantees that a search engine will never be forced to wait when requesting your sitemap. If you have enough content, search engines may hang up before your sitemap is generated, so this task is very useful.

### Generating sitemaps as static files

If you wish, you can generate a sitemap as a static file.

Just run this task:

```
node app apostrophe-site-map:map
```

When `--update-cache` is not given, this task generates an XML sitemap and displays it on the console. This is mostly useful for content strategy purposes. If your goal is to serve the sitemap to search engines, see above for a better way.

## How to tell Google about your sitemap

Create a `public/robots.txt` file if you do not already have one and add a Sitemap line. Here is a valid example for a site that doesn't have any other `robots.txt` rules:

```
Sitemap: http://EXAMPLE.com/sitemap.xml
```

You can also have other `robots.txt` directives if you wish.

On Google's next crawl of your site it should pick up on the presence of the sitemap.

## Changing the priority of pages and pieces

By default, an XML sitemap will assign a priority to a page based on its depth. The home page has a priority of 1.0 (the highest), a subpage of the home page 0.9, and so on.

Pieces receive a priority of 0.7; however if they have a `startDate` property (i.e. they are events) in the future, they bump up to 0.8, and if they have a `startDate` in the past they bump down to 0.6.

**You can also set the priority yourself.** Once you install this module you will discover that there is a new "sitemap priority" field in "page settings," and when editing a piece via the edit dialog box. You can set this field to any number between 0.0 and 1.0, with 1.0 being the highest.

As of this writing, Google suggests that they may use the priority to rank the importance of pages *relatively within your site.* **Please do not set all the priorities to 1.0. It will only hurt your chances of communicating which pages are most important to Google.**

## Content strategy

You can also use this module just to generate a map of your site for your own study:

```
node app apostrophe-site-map:map --format=text --indent
```

The result is a very informative depth-first list of pages. Note the use of leading spaces to indicate depth:

```
/
  /about
    /about/people
    /about/ducklings
/products
  /products/cheesemaker
```

You'll want to pipe that to a text file and consider printing it.

*The displayed "depth" of pieces won't always correspond directly to the pieces-pages that display them.* You might want to exclude them when generating content strategy maps.

## Warning: watch out for your custom stuff!

This module does the best it can.

It'll list your published pages, and your published pieces. And it'll rank future events higher than past events.

But it doesn't know anything about the custom URLs, independent of Apostrophe's usual mechanisms, that you're generating in your own creative and amazing modules.

If that's a concern for you, create `lib/modules/apostrophe-site-map/index.js` in your project, subclass the module, and override the `custom` method to output information about **additional** URLs. *Note: if you have multiple locales via `apostrophe-workflow` this method is called once per locale.* This method now receives `req, locale, callback` if written to accept three arguments.

It's straightforward: all you have to do is pass Apostrophe page objects, or anything else with an `_url` property and a `siteMapPriority` property, to `self.output`.

Here's a simple example. Note the use of `self.host` to get the "stem" of the URL (`http://mysite.com`).

For regular pages in the page tree, `level` starts at `0` (the home page) and increments from there for nested pages. For your own "pages," just keep that in mind. The higher the `level`, the lower the `priority` will be in the XML sitemap. Or pass the`siteMapPriority` property explicitly.

> This feature is **not** for changing priorities of existing pages and pieces. It is for your custom routes and dispatch URLs that the module cannot discover on its own. See the "page settings" dialog box or the edit dialog box for a field that lets you set the priority of an ordinary page or piece.

```javascript
// lib/modules/apostrophe-site-map/index.js, at project level, not in node_modules
module.exports = {
  construct: function(self, options) {
    self.custom = function(req, locale, callback) {
      // Discover something via the database, then...
      self.output({
        _url: 'http://mysite.com/myspecialplace',
        // Defaults to 0.5 if not set and a `level` property
        // cannot be used to infer it
        siteMapPriority: 0.9
      });
      return callback(null);
    };
  }
};
```

Note that `req` only has the same privileges as an anonymous site visitor. If you call `find` methods with it, you will only see what typical site visitors see. This is good, because **you don't want Google to index restricted pages.**

## How to exclude stuff

"I don't want thousands of blog posts in my sitemaps." OK, so do this in `app.js` when configuring the module:

Or do it in `app.js` when configuring the module:

```javascript
  {
    'apostrophe-site-map': {
      excludeTypes: [ 'apostrophe-blog-post' ]
    }
  }
```

You may specify multiple doc types to exclude. You may also exclude page types the same way by adding their doc type to the array, e.g., `styleguide`.

You can also do this at the command line, which is helpful when generating a map just for content strategy purposes:

```
node app apostrophe-site-map:map --format=text --indent --exclude-types=apostrophe-blog
```

Alternatively, you can set the `sitemap` option to `false` when configuring any module that extends `apostrophe-custom-pages` or `apostrophe-pieces`.

You can also explicitly set it to `true` if you wish to have sitemaps for a piece type that is normally excluded, like `apostrophe-users`. Of course this will only help if they have a `_url` property when fetched, usually via a corresponding module that extends `apostrophe-pieces-pages`.

## Removing the `siteMapPriority` field globally

You may wish to not include the `siteMapPriority` field on any pieces or pages. To do this, add a `noPriority` option set to `true` when configuring `apostrophe-site-map` in your `app.js`:

```javascript
  {
    'apostrophe-site-map': { noPriority: true }
  }
```

## Integration with the `apostrophe-workflow` module

If you are using the `apostrophe-workflow` module, the sitemap module will automatically fetch content for the live versions of all configured locales.


CHANGELOG.md
# Changelog

## 2.9.0 (2023-11-03)

Sitemaps now appear as XML source rather than awkwardly rendered text when viewed in the browser. There was never a problem with the XML from Google's perspective, but this improves the developer experience. Thanks to [norbie-nagy](https://github.com/norbie-nagy) for contributing this improvement.

## 2.8.0 (2021-02-22)

Adds new `getPageTree` method that returns the nested pages in the right order with pieces that are also pages. To be able to build a sitemap page from any project.

## 2.6.0 (2021-10-13)

Introduced a `rewriteUrl` method, which project developers can override to customize the URLs being output in the sitemap.

## 2.5.3

Disables the `siteMapPriority` schema field on the `trash` page to prevent an "unarranged field" warning about it.

## 2.5.1

Fixes a bug in the `exclude-types` argument usage.

## 2.5.0

Added a configuration option to remove `siteMapPriority` field globally

## 2.4.7

per Google's guidelines a `<link>` should contain an `<xhtml:link hreflang>` for every locale, including the locale of the `<link>` itself, which was formerly excluded. Thanks to Fredrik Ekelund for this contribution.

## 2.4.6

never generate a priority below 0.1.

## 2.4.5

clone the priority field before adding it so we do not get into issues with `arrangeFields`.

## 2.4.4

Fixes issue where children of unpublished top-level pages were left out of the site map. Additional documentation improvements.

## 2.4.3

fix for apps not using apostrophe-workflow, removing workflow-related xml tags that were left in the sitemap.

## 2.4.1

fix for static sitemap generation of workflow-driven sites without the `perLocale` option, along with new unit tests to verify this has no negative impact on "simple" sites. Also uses the newly exported `destroy` mechanism in its unit testing so we can use Mocha 5 and know that Apostrophe is truly freeing all resources in `apos.destroy`.

## 2.4.0

adds ability to exclude page types in addition to piece types.

## 2.3.3

a bug that broke static sitemap file generation in the absence of workflow was fixed.

## 2.3.2

a bug that broke the sitemap generator in the absence of workflow was fixed. Thanks to Peter Shaw.

## 2.3.1

a bug that broke the sitemap generator in the presence of private workflow locales was fixed. Thanks to Albert R. Timashev.

## 2.3.0

sitemaps for sites localized with the `apostrophe-workflow` module now include pointers to alternate language versions of each document.

## 2.2.1

* The command line `node app apostrophe-site-map:map --update-cache` can be used to update the sitemap that will be sent from Apostrophe's internal cache without waiting for the cache to expire. If the task is scheduled to run more often then once an hour, then a search engine will never be asked to wait a long time to generate it. For sites with many pages and pieces this can be critical.

## 2.2.0

* `piecesPerBatch` option for performance. Still defaults to processing 100 pieces at a time.
* Support for the `hostname` option of `apostrophe-workflow`.

## 2.1.1

short-lived bug affecting command line tasks.

## 2.1.0

sitemaps are now served dynamically. They are stored in Apostrophe's cache for a configurable period of time. There is no need to run a command line task, or mess around with static files. Please note that you must remove existing static sitemap files first. See the documentation for important recommendations. Thanks to Michelin for their support of this work.

The documentation has also been overhauled thoroughly to be completely accurate for Apostrophe 2.x.

## 2.0.4

workflow-aware; new features providing compatibility with the apostrophe-workflow module.

## 2.0.3

documentation updates.

## 2.0.1-2.0.2

minor bug fixes.

## 2.0.0

initial port to Apostrophe 2.x.

## 2.2.0

enhance performances with many pieces


index.js
var moment = require('moment');
var async = require('async');
var _ = require('@sailshq/lodash');
var fs = require('fs');
var url = require('url');

var defaultLocale = 'default';

module.exports = {

  // Cache sitemaps for 1 hour by default. Depending on pagerank
  // Google may look at your sitemap somewhere between daily and
  // monthly, so don't get your hopes up too far about changing this

  cacheLifetime: 60 * 60,

  piecesPerBatch: 100,

  moogBundle: {
    modules: [ 'apostrophe-site-map-custom-pages', 'apostrophe-site-map-pieces' ],
    directory: 'lib/modules'
  },

  afterConstruct: function(self) {
    self.apos.tasks.add(self.__meta.name, 'map', self.mapTask);
    self.apos.tasks.add(self.__meta.name, 'clear', self.clearTask);
    self.addRoutes();
    self.enableCache();
  },

  construct: function(self, options) {

    self.caching = true;

    self.cacheLifetime = options.cacheLifetime;

    self.piecesPerBatch = options.piecesPerBatch;

    self.baseUrl = options.baseUrl || self.apos.baseUrl;

    self.clearTask = function(apos, argv, callback) {
      // Just forget the current sitemaps to make room
      // for regeneration on the next request
      return self.cache.clear(callback);
    };

    self.mapTask = function(apos, argv, callback) {
      if (argv['update-cache']) {
        self.caching = true;
      } else {
        self.caching = false;
      }

      if (!self.baseUrl) {
        return callback(new Error(
          'You must specify the top-level baseUrl option when configuring Apostrophe\n' +
          'to use this task. Example: baseUrl: "http://mycompany.com"\n\n' +
          'Note there is NO TRAILING SLASH.\n\n' +
          'Usually you will only do this in data/local.js, on production.'
        ));
      }

      return self.map(callback);
    };

    self.map = function(callback) {

      self.workflow = self.apos.modules['apostrophe-workflow'];

      var argv = self.apos.argv;

      if (self.caching) {
        self.cacheOutput = [];
      }
      return async.series([
        lock,
        init,
        map,
        hreflang,
        write,
        unlock
      ], callback);

      function lock(callback) {
        return self.apos.locks.lock('apostrophe-site-map', callback);
      }

      function init(callback) {
        self.format = argv.format || options.format || 'xml';

        self.indent = (typeof(argv.indent) !== 'undefined') ? argv.indent : options.indent;

        self.excludeTypes = options.excludeTypes || [];

        if (argv['exclude-types']) {
          self.excludeTypes = self.excludeTypes.concat(argv['exclude-types'].split(','));
        }

        self.perLocale = options.perLocale || argv['per-locale'];
        // Exception: plaintext sitemaps and sitemap indexes don't go
        // together, so we can presume that if they explicitly ask
        // for plaintext they are just doing content strategy and we
        // should produce a single report
        if (self.format === 'text') {
          self.perLocale = false;
        }
        return callback(null);
      }

      function map(callback) {
        self.maps = {};
        self.today = moment().format('YYYY-MM-DD');

        var locales = [ defaultLocale ];

        if (self.workflow) {
          locales = _.filter(_.keys(self.workflow.locales), function(locale) {
            return !locale.match(/-draft$/) && !self.workflow.locales[locale].private;
          });
        }

        return async.eachSeries(locales, function(locale, callback) {
          var req = self.apos.tasks.getAnonReq({ locale: locale });
          req.locale = locale;
          return async.series([
            _.partial(self.getPages, req, locale),
            _.partial(self.getPieces, req, locale),
            function(callback) {
              if (self.custom.length === 1) {
                return self.custom(callback);
              } else {
                return self.custom(req, locale, callback);
              }
            }
          ], callback);
        }, function(err) {
          if (err) {
            return callback(err);
          }
          return callback(null);
        });
      }

      function hreflang(callback) {

        var alternativesByGuid = {};

        each(function(entry) {
          if (!alternativesByGuid[entry.url.workflowGuid]) {
            alternativesByGuid[entry.url.workflowGuid] = [];
          }
          alternativesByGuid[entry.url.workflowGuid].push(entry);
        });

        each(function(entry) {
          if (self.workflow) {
            entry.url['xhtml:link'] = [{
              _attributes: {
                rel: 'alternate',
                hreflang: entry.url.workflowLocale,
                href: entry.url.loc
              }
            }];
          }
          var alternatives = alternativesByGuid[entry.url.workflowGuid];
          _.each(alternatives, function(alternative) {
            if (alternative === entry) {
              return;
            }
            entry.url['xhtml:link'].push({
              _attributes: {
                rel: 'alternate',
                hreflang: alternative.url.workflowLocale,
                href: alternative.url.loc
              }
            });
          });
        });

        each(function(entry) {
          delete entry.url.workflowLocale;
          delete entry.url.workflowGuid;
        }, true);

        return setImmediate(callback);

        function each(iterator, ignoreWorkflow) {
          _.each(self.maps, function(map) {
            _.each(map, function(entry) {
              if (typeof(entry) !== 'object') {
                return;
              }

              if (!entry.url.workflowGuid && !ignoreWorkflow) {
                return;
              }
              iterator(entry);
            });
          });
        }

      }

      function write(callback) {
        return self.writeSitemap(callback);
      }

      function unlock(callback) {
        return self.apos.locks.unlock('apostrophe-site-map', callback);
      }
    };

    self.getPages = function(req, locale, callback) {
      return self.apos.pages.find(req).areas(false).joins(false).sort({ level: 1, rank: 1 }).toArray(function(err, pages) {
        if (err) {
          return callback(err);
        }
        _.each(pages, self.output);
        return callback(null);
      });
    };

    self.getPieces = function(req, locale, callback) {
      const modules = self.getPiecesModules();

      return async.eachSeries(modules, function(module, callback) {
        if (_.includes(self.excludeTypes, module.name)) {
          return setImmediate(callback);
        }

        // Paginate through 100 (by default) at a time to
        // avoid slamming memory
        var done = false;
        var skip = 0;
        return async.whilst(
          function() { return !done; },
          function(callback) {

          return self.findPieces(req, module)
            .skip(skip)
            .limit(self.piecesPerBatch)
            .toArray(function(err, pieces) {
              _.each(pieces, function(piece) {
                if (!piece._url) {
                  // This one has no page to be viewed on
                  return;
                }

                // Results in a reasonable priority relative
                // to regular pages
                piece.level = 3;
                // Future events are interesting,
                // past events are boring
                if (piece.startDate) {
                  if (piece.startDate > self.today) {
                    piece.level--;
                  } else {
                    piece.level++;
                  }
                }
                self.output(piece);
              });
              if (!pieces.length) {
                done = true;
              } else {
                skip += pieces.length;
              }
              return callback(null);
          });
        }, callback);
      }, callback);
    };

    self.getPiecesModules = () => {
      return Object.values(self.apos.modules).filter((mod) => {
        return mod.__meta.chain.some((meta) => meta.name === 'apostrophe-pieces')
      })
    }

    self.writeSitemap = function(callback) {
      if (!self.perLocale) {
        // Simple single-file sitemap
        self.file = self.caching ? 'sitemap.xml' : (self.apos.argv.file || '/dev/stdout');
        var map = _.map(_.keys(self.maps), function(locale) {
          return _.map(self.maps[locale], self.stringify).join('\n');
        }).join('\n');
        self.writeMap(self.file, map);
      } else {
        // They should be broken down by host,
        // in which case we automatically place them
        // in public/sitemaps in a certain naming pattern
        self.ensureDir('sitemaps');
        _.each(self.maps, function(map, key) {
          var extension = (self.format === 'xml') ? 'xml' : 'txt';
          map = _.map(map, self.stringify).join('\n');
          self.writeMap('sitemaps/' + key + '.' + extension, map);
        });
        self.writeIndex();
      }
      if (self.caching) {
        return self.writeToCache(callback);
      }
      return callback(null);
    };

    // If `value` is not an object, it is returned as-is,
    // or with < & > escaped if `self.format` is `xml`.
    //
    // If it is an object, it is converted to XML elements,
    // one for each property; they may have sub-elements if
    // the properties contain objects. The _attributes
    // property is converted to attributes. Array
    // properties are converted to a series of elements.
    //
    // TODO: this is clearly yak-shaving, but the data format
    // is nice. See if there's another library that takes the same
    // or substantially the same format.

    self.stringify = function(value) {
      if (Array.isArray(value) && (self.format !== 'xml')) {
        return value.join('');
      }
      if (typeof(value) !== 'object') {
        if (self.format === 'xml') {
          return self.apos.utils.escapeHtml(value);
        }
        return value;
      }
      var xml = '';
      _.each(value, function(v, k) {
        if (k === '_attributes') {
          return;
        }
        if (Array.isArray(v)) {
          _.each(v, function(el) {
            element(k, el);
          });
        } else {
          element(k, v);
        }
        function element(k, v) {
          xml += '<' + k;
          if (v && v._attributes) {
            _.each(v._attributes, function(av, a) {
              xml += ' ' + a + '="' + self.apos.utils.escapeHtml(av) + '"';
            });
          }
          xml += '>';
          xml += self.stringify(v || '');
          xml += '</' + k + '>\n';
        }
      });
      return xml;
    };

    self.ensureDir = function(dir) {
      if (!self.caching) {
        dir = self.apos.rootDir + '/public/' + dir;
        try {
          fs.mkdirSync(dir);
        } catch (e) {
          // exists
        }
      }
    };

    self.writeIndex = function() {
      var now = new Date();
      if (!self.baseUrl) {
        throw new Error(
          'You must specify the top-level baseUrl option when configuring Apostrophe\n' +
          'to use sitemap indexes. Example: baseUrl: "http://mycompany.com"\n\n' +
          'Note there is NO TRAILING SLASH.\n\n' +
          'Usually you will override this in data/local.js, on production.'
        );
      }
      self.writeFile('sitemaps/index.xml',

      '<?xml version="1.0" encoding="UTF-8"?>\n' +
        '<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"' +
        ' xmlns:xhtml="http://www.w3.org/1999/xhtml">\n' +
        _.map(_.keys(self.maps), function(key) {
          var map = self.maps[key];
          var sitemap = '  <sitemap>\n' +
            '    <loc>' + self.baseUrl + self.apos.prefix + '/sitemaps/' + key + '.xml'
              + '</loc>\n' +
            '    <lastmod>' + now.toISOString() + '</lastmod>\n' +
          '  </sitemap>\n';
          return sitemap;
        }).join('') +
        '</sitemapindex>\n'
      );

    };

    self.writeMap = function(file, map) {
      if (self.format === 'xml') {
        self.writeXmlMap(file, map);
      } else {
        self.writeFile(file, map);
      }
    };

    self.writeXmlMap = function(file, map) {
      self.writeFile( file,
        '<?xml version="1.0" encoding="UTF-8"?>\n' +
        '<urlset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" ' +
        'xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 ' +
        'http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd ' +
        'http://www.w3.org/TR/xhtml11/xhtml11_schema.html ' +
        'http://www.w3.org/2002/08/xhtml/xhtml1-strict.xsd" ' +
        'xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" ' +
        'xmlns:xhtml="http://www.w3.org/TR/xhtml11/xhtml11_schema.html">\n' +
        map +
        '</urlset>\n'
      );
    };

    self.writeFile = function(filename, s) {
      if (!self.caching) {
        filename = require('path').resolve(self.apos.rootDir + '/public', filename);
        if (filename === '/dev/stdout') {
          // Strange bug on MacOS when using writeFileSync with /dev/stdout
          fs.writeSync(1, s);
        } else {
          fs.writeFileSync(filename, s);
        }
      } else {
        self.cacheOutput.push({
          filename: filename,
          data: s,
          createdAt: new Date()
        });
      }
    };

    self.writeToCache = function(callback) {
      return async.series([
        clear,
        insert
      ], callback);

      function clear(callback) {
        return self.cache.clear(callback);
      }

      function insert(callback) {
        return async.eachSeries(self.cacheOutput, function(doc, callback) {
          return self.cache.set(doc.filename, doc, self.cacheLifetime, callback);
        }, callback);
      }
    };

    // Override to do more. You can invoke `self.output(doc)`
    // from here as many times as you like.

    self.custom = function(req, locale, callback) {
      return callback(null);
    };

    self.findPieces = function(req, module, projection = {}) {
      return module.find(req, {}, projection).published(true).joins(false).areas(false);
    };

    // Output the sitemap entry for the given doc, including its children if any.
    // The entry is buffered for output as part of the map for the appropriate
    // locale. If the workflow module is not in use they all accumulate together
    // for a "default" locale. Content not subject to workflow is grouped with
    // the "default" locale. If workflow is active and the locale is not configured
    // or is marked private, the output is discarded.

    self.output = function(page) {
      var locale = page.workflowLocale || defaultLocale;
      if (self.workflow) {
        if (!self.workflow.locales[locale]) {
          return;
        }
        if (self.workflow.locales[locale].private) {
          return;
        }
      }

      if (!_.includes(self.excludeTypes, page.type)) {
        var url;

        if (self.format === 'text') {
          if (self.indent) {
            var i;

            for (i = 0; (i < page.level); i++) {
              self.write(locale, '  ');
            }

            self.write(locale, self.rewriteUrl(page._url) + '\n');
          }
        } else {
          url = self.rewriteUrl(page._url);
          var priority = (page.level < 10) ? (1.0 - page.level / 10) : 0.1;

          if (typeof (page.siteMapPriority) === 'number') {
            priority = page.siteMapPriority;
          }

          self.write(locale, {
            url: {
              priority: priority,
              changefreq: 'daily',
              loc: url,
              workflowGuid: page.workflowGuid,
              workflowLocale: locale
            }
          });
        }
      }

    };

    // Append `s` to an array set aside for the map entries
    // for the host `locale`.

    self.write = function(locale, s) {
      self.maps[locale] = self.maps[locale] || [];
      self.maps[locale].push(s);
    };

    self.addRoutes = function() {
      // Deliver from our tiny little fake cache filesystem
      self.apos.app.get('/sitemap.xml', function(req, res) {
        return self.sendCache(res, 'sitemap.xml');
      });
      self.apos.app.get('/sitemaps/*', function(req, res) {
        return self.sendCache(res, 'sitemaps/' + req.params[0]);
      });
    };

    self.sendCache = function(res, path) {
      return self.cache.get(path, function(err, file) {
        if (err) {
          return fail(err);
        }
        if (!file) {
          // If anything else exists in our little filesystem, this
          // should be a 404 (think of a URL like /sitemap/madeupstuff).
          // Otherwise it just means the
          // cache has expired or has never been populated.
          //
          // Check for the sitemap index or, if we're not
          // running in that mode, check for sitemap.xml
          //
          // Without this check every 404 would cause a lot of work to be done.
          return self.cache.get(self.perLocale ? 'sitemaps/index.xml' : 'sitemap.xml', function(err, exists) {
            if (err) {
              return fail(err);
            }
            if (exists) {
              return notFound();
            }
            return self.cacheAndRetry(res, path);
          });
        }
        return res.contentType('text/xml').send(file.data);
      });

      function notFound() {
        return res.status(404).send('not found');
      }

      function fail(err) {
        console.error(err);
        return res.status(500).send('error');
      }
    };

    self.cacheAndRetry = function(res, path) {
      return self.map(function(err) {
        if (err) {
          return fail(err);
        }
        return self.sendCache(res, path);
      });
      function fail(err) {
        console.error(err);
        return res.status(500).send('error');
      }
    };

    self.enableCache = function() {
      self.cache = self.apos.caches.get('apostrophe-sitemap');
    };

    // Override this method at project level to customize the URLs output in the sitemap.
    // Useful in headless applications where the URLs visible to Apostrophe,
    // acting as a backend, differ from those visible to the public

    self.rewriteUrl = url => {
      return url;
    };

    self.getPageTree = async (req) => {
      const excludedTypes = [
        'workflow-document',
        ...self.options.excludeTypes || [],
        ...self.options.excludeTypesFromPageTree || [],
      ]

      const pages = await getPages();
      const pagesWithPieces = await getPieces(pages);

      return rewriteUrls(pagesWithPieces);

      async function getPages () {
        try {
          const projection =  {
            _id: 1,
            title: 1,
            type: 1,
            slug: 1,
            path: 1,
            rank: 1,
            level: 1
          }

          const criteria = {
            type: {
              $nin: excludedTypes
            }
          }

          const homePage = await self.apos.pages
          .find(req, { level: 0 }, projection)
          .children({ depth: 1000, orphan: null, projection, and: criteria })
          .toObject();

          return [
            {
              ...homePage,
              _children: []
            },
            ...homePage._children
          ]
        } catch (err) {
          self.apos.utils.error(err);
        }
      }

      async function getPieces (pages) {
        const piecesModules = self.getPiecesModules();
        const pieces = [];

        for (const mod of piecesModules) {
          if (excludedTypes.includes(mod.name)) {
            continue;
          }

          await fetchPieces(req, {
            mod,
            skip: 0,
            pieces
          });
        }

        return insertPieces(pages, pieces)

        async function fetchPieces (req, {
          mod, skip, pieces
        }) {
          try {
            const fetchedPieces = await self.findPieces(req, mod, { _id: 1, title: 1, _url: 1 })
              .skip(skip)
              .limit(self.piecesPerBatch)
              .toArray();

            if (!Array.isArray(fetchedPieces)) {
              return;
            }

            fetchedPieces.forEach(piece => {
              if (piece._url && !excludedTypes.includes(piece.type)) {
                pieces.push(piece);
              }
            });

            if (fetchedPieces.length) {
              await fetchPieces(req, {
                mod,
                skip: skip + fetchedPieces.length,
                pieces
              });
            }
          } catch (err) {
            self.apos.utils.error(err);
          }
        }

        function insertPieces (pages, pieces) {
          return pages.reduce((acc, page) => {
            const filledChildren = page._children.length
              ? insertPieces(page._children, pieces)
              : page._children

            const childrenPieces = pieces
              .filter((piece) => piece._parentUrl === page._url)

            return [
              ...acc,
              {
                ...page,
                _children: [
                  ...filledChildren,
                  ...childrenPieces
                ]
              }
            ]
          }, [])
        }
      }

      function rewriteUrls (pages = []) {
        return pages.reduce((acc, page) => {
          return [
            ...acc,
            {
              ...page,
              _url: self.rewriteUrl(page._url),
              _children: rewriteUrls(page._children)
            }
          ]
        }, [])
      }
    };
  }
};
